### DISCUSSION TITLE
Embedded objects with ts-morph error

### DISCUSSION THREAD

**Discussion by zhalil** [mikro-orm/mikro-orm#6134] [General]
### Describe the bug

hello! i have some problems with embedded properties using ts-morph & nestjs , the documentation shows the following example
```ts
import { Embeddable, Embedded, Entity, PrimaryKey, Property } from '@mikro-orm/core';

@Embeddable()
export class Address {

  @Property()
  street!: string;

  @Property()
  postalCode!: string;

  @Property()
  city!: string;

  @Property()
  country!: string;

}

@Entity()
export class User {

  @PrimaryKey()
  id!: number;

  @Embedded()
  address!: Address;

}
```
but when I try to repeat the same thing I get an error if i run migration create, my source:
```ts

import { Embedded, Entity, PrimaryKey } from "@mikro-orm/core";
import { TelegramId } from "../vo/telegram-id.vo";


@Entity()
export class User {
    @PrimaryKey({
        columnType: "uuid"
    })
    id!: string;

    @Embedded()
    private telegramId!: TelegramId;
    
}
```

```ts

import { Embeddable, Embedded, Property } from "@mikro-orm/core";
@Embeddable()
export class TelegramId {
  @Property({
    name: "telegram_id"
  })
  private value: number;
}
```

errors : 
MetadataError: Entity 'TelegramId' was not discovered, please make sure to provide it in 'entities' array when initializing the ORM (used in User.telegramId)

but if i use : 

```ts
@Embedded(() => TelegramId)
 private telegramId!: TelegramId;
```
everything works

my mikro config:
```ts
import { Migrator } from '@mikro-orm/migrations';
import {defineConfig, PostgreSqlDriver} from "@mikro-orm/postgresql";
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';
export default defineConfig({
    entities: ['dist/src/*/domain/**/*/*.entity.js'],
    entitiesTs: ['src/**/domain/*/*.entity.ts'],
    host: 'postgres',
    user:"app",
    dbName: 'app',
    password: "pgapppassword",
    driver: PostgreSqlDriver,
    port: 5432,
    extensions: [Migrator],
    metadataProvider: TsMorphMetadataProvider,
    migrations: {
        path: './db/migrations',
    }
})
```

### Reproduction

https://github.com/zhalil/microtest

location config : root
location entity: ./src/user/domain/entity
embeddable entity location : ./src/user/domain/vo

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.13

### Node.js version

22.9.0

### Operating system

docker node:22.9.0-alpine

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
ts-morph is a metadata provider, its output is a string name, so equivalent of doing this

```ts
@Embedded('TelegramId')
 private telegramId!: TelegramId;
```

you still need to make sure that embeddable is discovered, which can happen automatically only if you provide the class reference yourself, as in `@Embedded(() => TelegramId)`.

your embeddables are defined in a separate files that don't match the pattern you have in your ORM config, so they wont be discovered directly. 

try adding this to the ORM config to include them:

```ts
entities: ['dist/src/*/domain/**/*/*.{entity|vo}.js'],
entitiesTs: ['src/**/domain/*/*.{entity|vo}.ts'],
```

note that `dist/src` is quite suspicious and usually means some issue in your setup, you should have the files compiled directly into `dist` folder instead of nesting into `src`, but that's another story.

**Reply by zhalil** [reply]
oh, now I understand how it works, thank you very much!

**Reply by zhalil** [reply]
I tried to do it as you said, but even if I make telegram-id.entity.ts and move it to the same folder where the entities are, I get the error in runtime
"
ERROR [ExceptionHandler] Entity 'any' was not discovered, please make sure to provide it in 'entities' array when initializing the ORM (used in User.telegramId)
"
current code:
```ts
import { Embedded, Entity, PrimaryKey } from "@mikro-orm/core";
import {TelegramId} from "./telegram-id.entity";


@Entity()
export class User {
    @PrimaryKey({
        columnType: "uuid"
    })
    id!: string;

    @Embedded()
    private telegramId!: TelegramId;

}
```

```ts
   entities: ['./dist/src/*/**/***/**/*.entity.js',],
    entitiesTs: ['./src/*/**/***/**/*.entity.ts',],
```

```ts
  imports: [
    MikroOrmModule.forFeature([User, TelegramId])
  ],
```

**Reply by zhalil** [reply]
also a question, is it possible to make embeddable primaries key?

**Reply by B4nan** [reply]
please dont reopen discussions just to ask something, you just sent me two notifications for no reason.

> also a question, is it possible to make embeddable primaries key?

nope, for PKs you'd have to use a custom type

**Reply by zhalil** [reply]
> Я попытался сделать так, как вы сказали, но даже если я создам telegram-id.entity.ts и перемещу его в ту же папку, где находятся сущности, во время выполнения возникает ошибка « ERROR [ExceptionHandler] Entity 'any' не обнаружена, пожалуйста, обязательно укажите ее в массиве 'entities' при инициализации ORM (используется в User.telegramId) » текущий код:
> 
> ```ts
> import { Embedded, Entity, PrimaryKey } from "@mikro-orm/core";
> import {TelegramId} from "./telegram-id.entity";
> 
> 
> @Entity()
> export class User {
>     @PrimaryKey({
>         columnType: "uuid"
>     })
>     id!: string;
> 
>     @Embedded()
>     private telegramId!: TelegramId;
> 
> }
> ```
> 
> ```ts
>    entities: ['./dist/src/*/**/***/**/*.entity.js',],
>     entitiesTs: ['./src/*/**/***/**/*.entity.ts',],
> ```
> 
> ```ts
>   imports: [
>     MikroOrmModule.forFeature([User, TelegramId])
>   ],
> ```

migrations were created successfully, but the runtime problem remains

**Reply by zhalil** [reply]
> пожалуйста, не открывайте обсуждения снова, просто чтобы что-то спросить. Вы только что отправили мне два уведомления без причины.
> 
> > еще вопрос, можно ли сделать встраиваемые первичные ключи?
> 
> нет, для PK вам придется использовать пользовательский тип

my mistake, sorry, I rarely create issues, I'll know for the future, but what could be the problem? the config is the same but it works for migrations but not for runtime

**Reply by B4nan** [reply]
I would expect the very opposite, your `forFeature` call can help with runtime (but only if you use `autoLoadEntities` option of the nest adapter), while it has no effect on the ORM CLI.

**Reply by zhalil** [reply]
> I would expect the very opposite, your `forFeature` call can help with runtime (but only if you use `autoLoadEntities` option of the nest adapter), while it has no effect on the ORM CLI.

I tried to simply add entity autoloading to the config
```ts
    MikroOrmModule.forRoot({...config, autoLoadEntities: true}),
```
and got an error:
 ERROR [ExceptionHandler] Duplicate entity names are not allowed: User, TelegramId

after that I tried to split the config:
```ts
export default defineConfig({
    host: 'postgres',
    user:"app",
    dbName: 'app',
    password: "pgapppassword",
    driver: PostgreSqlDriver,
    port: 5432,
    extensions: [Migrator],
    metadataProvider: TsMorphMetadataProvider,
    migrations: {
        path: './db/migrations',
    }
});

MikroOrmModule.forRoot({...baseConfig, autoLoadEntities: true}),
```
but now i'm back to the previous mistake:
ERROR [ExceptionHandler] Entity 'any' was not discovered, please make sure to provide it in 'entities' array when initializing the ORM (used in User.telegramId)

**Reply by B4nan** [reply]
> ERROR [ExceptionHandler] Duplicate entity names are not allowed: User, TelegramId

This means you have multiple different entities with the same name but different prototype. That can happen if you mess up with the folder based discovery, namely when you discover JS entities inside TS context.

> ERROR [ExceptionHandler] Entity 'any' was not discovered, please make sure to provide it in 'entities' array when initializing the ORM (used in User.telegramId)

And this is most likely an import cycle, resulting in `any` type.

Try to look at https://github.com/mikro-orm/nestjs-realworld-example-app for inspiration.
